keys.go

```diff
@@ -0,0 +1,70 @@
+package paseto
```

Comment:
If possible without introducing dependency cycles, I think we can probably contain most of PASREK in its own submodule? (As opposed to including it directly in the main `paseto` package)?

Reply:
I have tried to do it, after getting deeper in the paserk standard, and I see the following possibilities:

1. Use paserk package for defining encoders/decoders, and use the binary (`[]byte`) methods to export keys/create new ones. In this case paseto package would depend on paserk. An example (currently without error handling) would look like:
    ```golang
    key, err := paseto.V2SymmetricKeyFromHex("707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f")
    fmt.Println(key.ExportHex())
    password, memlimit, opslimit := "correct horse battery staple", 268435456, 3
    paserkStr, err := key.ToPaserk(paserk.EncodeWithPassword(password, memlimit, opslimit))
    fmt.Println(paserkStr)
    decodedKey, err := paseto.V2SymmetricKeyFromPaserk(paserkStr, paserk.DecodeWithPassword(password, memlimit, opslimit))
    fmt.Println(decodedKey.ExportHex())
    // Output:
    // 707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f
    // k2.local-pw.pO9bFwcQ4EvMgi05yZUZYAAAAAAQAAAAAAAAAwAAAAHQDD7woll221xBostsekagJ31WkLjCKzV1emeOK91FbTeyY-NRZAqnoLnGHs30MEbbL7SVqmtPhELmR8ZbrYIPOu6LJh2IQ4bs-xjkPk6UB0t5E9RcYhEt
    // 707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f
    ```
    For this `v2_keys.go` needs to be extended as follows:
    ```golang
    func (k V2SymmetricKey) ToPaserk(encoding paserk.Encoding) (string, error) {
        // we can simply pass the key to paserk as the key has a public ExportHex method and a Key interface is defined in paserk
        return encoding.Encode(&k)
    }

    func V2SymmetricKeyFromPaserk(paserkStr string, encoding paserk.Encoding) (V2SymmetricKey, error) {
        // Decoder can't return a V2SymmetricKey, as paserk can't use types from paseto package.
        // I can't se a way to hide it with an interface.
        data, err := encoding.Decode(paserkStr, paserk.KeyPurposeLocal, paserk.KeyVersionV2)
        if err != nil {
            // even though we return error, return a random key here rather than
            // a nil key
            return NewV2SymmetricKey(), errors.New("Key incorrect length")
        }
        return V2SymmetricKeyFromBytes(bytes)
    }
    ```
1. The previous solution may be implemented with even more specific encoders like:
    ```golang
    paserkStr, err = key.ToPaserk(paserk.V2EncodeLocalWithPassword(password, memlimit, opslimit))
    decodedKey, err := paseto.V2SymmetricKeyFromPaserk(paserkStr, paserk.V2DecodeLocalWithPassword(password, memlimit, opslimit))
    ```
    This may be useful for sealing, where encoder params depend on key version (ie. NIST vs ed curve keys).
1. Reverse the direction (ie. paserk package would depend on paseto)
    ```golang
    key, err := paseto.V2SymmetricKeyFromHex("707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f")
    fmt.Println(key.ExportHex())
    password, memlimit, opslimit := "correct horse battery staple", 268435456, 3
    encoder = paserk.V2PasswordEncoder(password, memlimit, opslimit)
    paserkStr, err := encoder.Encode(key)
    fmt.Println(paserkStr)
    decoder = paserk.V2PasswordDecoder(password, memlimit, opslimit)
    decodedKey, err := decoder.Decode(paserkStr)
    fmt.Println(decodedKey.ExportHex())
    // Output:
    // 707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f
    // k2.local-pw.pO9bFwcQ4EvMgi05yZUZYAAAAAAQAAAAAAAAAwAAAAHQDD7woll221xBostsekagJ31WkLjCKzV1emeOK91FbTeyY-NRZAqnoLnGHs30MEbbL7SVqmtPhELmR8ZbrYIPOu6LJh2IQ4bs-xjkPk6UB0t5E9RcYhEt
    // 707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f    ```
    The latter seems to provide a cleaner API, and fits better to your core paseto implementation.

I would plan to implement the Marshaler/Unmarshaler interface to call the raw paserk encoding of the keys. Such a function (like `func (k *V2SymmetricKey) UnmarshalJSON(jsonstr []byte) error`) must be implemented in the paseto package because of the receiver. It can only be done of paseto package depends on paserk and not vica versa.

So which option would you prefer? Or can you come up with a different idea?

Reply:
Architecturally I think it probably makes the most sense for the paserk package to depend upon paseto (as in option 3). If you're using paserk, you'll be using it for paseto (but if you're using paseto, you might not use paserk). This would also matches the way the reference implementation is setup (not the only reason to do it, but makes things a bit simpler if we want to compare implementations).

I do also like the API simplicity of using paserk to first construct a key, and then using paseto to sign/encrypt the message. There's a nice separation of concerns there, and means that we can keep the key export options within the paseto package quite simple.